### PR TITLE
HPO file format change

### PR DESCRIPTION
**Pull request by dnil**
This PR adds a functionality or fixes a bug.

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by northwestwitch** [comment]
> With a fresh brain I may try to do something a bit more clever with the phenotype annotations to fit better with the genes and diseases we have loaded, and maybe add another more overall test on it.

Do you want to add the things you mentioned or it's fine to review?

File: scout/commands/setup/setup_scout.py [hunk 1]
```diff
@@ -92,6 +92,14 @@ def abort_if_false(ctx, param, value):
         "phenotype_to_genes.txt"
     ),
 )
+@click.option(
+    "--hpo-phenotype-annotation",
+    type=click.Path(exists=True),
+    help=(
```
Comment (northwestwitch) [inline scout/commands/setup/setup_scout.py:98]: Question: don't we need to update also the command line's `scout update hpo` with this new param? In the situation you downloaded the phenotype.hpoa and you want to use it when running the command scout update hpo?

File: scout/load/hpo.py [hunk 2]
```diff
@@ -6,10 +6,14 @@
```
Comment (northwestwitch) [inline scout/load/hpo.py:6]: The pp import above can be removed

**Review by northwestwitch — COMMENTED** [review]
Very nice fix and simplified code! 

But the new file can't be used from the command line (scout update hpo), or?

image

File: scout/load/hpo.py [hunk 3]
```diff
@@ -136,31 +150,49 @@ def load_disease_terms(adapter, genemap_lines, genes=None, hpo_disease_lines=Non
...
-            if hpo_terms:
-                disease_info["hpo_terms"] = hpo_terms
+        if disease_id in disease_annotations:
+            if "hpo_terms" in disease_info:
+                disease_info["hpo_terms"].update(disease_annotations[disease_id]["hpo_terms"])
+            else:
+                disease_info["hpo_terms"] = set(disease_annotations[disease_id]["hpo_terms"])
+
+            for hpo_term in disease_info["hpo_terms"]:
+                if hpo_term in hpo_term_to_symbol:
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/load/hpo.py]: I had already joined them based on the "code smell" suggestion, but I like this better - easier to read!

[the remark this answers is not in the document]
Comment (dnil) [reply scout/load/hpo.py]: Also helped see an additional issue with it!

File: scout/commands/setup/setup_scout.py [hunk 4]
```diff
@@ -92,6 +92,14 @@ def abort_if_false(ctx, param, value):
         "phenotype_to_genes.txt"
     ),
 )
+@click.option(
+    "--hpo-phenotype-annotation",
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/commands/setup/setup_scout.py:96]: Indeed - thought the same. One could also actually just add an `-f` option for a folder, but it gets a little complex to keep both ways.

File: scout/build/disease.py [hunk 5]
```diff
@@ -71,6 +71,7 @@ def build_disease_term(disease_info, alias_genes={}):
                     hgnc_ids.add(hgnc_id)
         else:
             LOG.debug("Gene symbol %s could not be found in database", hgnc_symbol)
+            hgnc_symbols_not_found.add(hgnc_symbol)
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/build/disease.py:74]: Ah, thanks - probably in an envisioned summary warning message!

File: scout/load/hpo.py (hunk 2 above)
Comment (dnil) [reply scout/load/hpo.py:6]: Already done! 😁

File: scout/commands/setup/setup_scout.py (hunk 1 above)
Comment (dnil) [reply scout/commands/setup/setup_scout.py:98]: If I recall correctly (hey, it was last week after all) we only need the flag for Diseases, and then of course when we update everything. 

The ones that change is the ones that used to call `parse_hpo_diseases`, and that would include `scout update diseases` but not `scout update hpo`. 

But, this is a little tricky so I will reread it again - they have layered fallback to default files if one is missing, so its easy to miss..

Comment (dnil) [reply scout/commands/setup/setup_scout.py:98]: Ok, no, I stand by that.

update hpo (https://github.com/Clinical-Genomics/scout/blob/e48092a9aaa90f85545b1dd8363c7049799a73e2/scout/commands/update/hpo.py#L54)
will call `load_hpo_terms`, (https://github.com/Clinical-Genomics/scout/blob/e48092a9aaa90f85545b1dd8363c7049799a73e2/scout/load/hpo.py#L58) which never loads diseases. 

This is functionally in contrast to `load_hpo`, which is only called from the `setup` command, and will load both HPO and Diseases on its own rather than calling the two other functions, for some awkward historical reason or maybe optimisation reasons. But I'm not about to change that today. We could put harmonising that on that major list.

File: scout/commands/update/genes.py [hunk 6]
```diff
@@ -12,9 +12,6 @@
 The hgnc file will determine which genes that are added and most of the meta information.
 The ensembl gene file will add coordinates and the exac file will add pLi scores.
 
-Created by Måns Magnusson on 2015-01-14.
-Copyright (c) 2015 __MoonsoInc__. All rights reserved.
```
[the remark this answers is not in the document]
Comment (dnil) [reply scout/commands/update/genes.py:16]: Indeed! 😁

I kind of doubt the company did ever exist.. But, for formality, any such history will still be on the repo, and the full code still has its common 2014 copyright statement in the MIT license statement, Måns included. The one referenced in the MIT license file is also the only copyright statement that we are formally obliged to keep.

[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/commands/update/genes.py:16]: @moonso

**Comment by northwestwitch** [comment]
> No, its for `scout update diseases` only. The file format change only dislodged the disease info from two of the files, the rest is still there.

AHH OK, it's confusing with all these files involved!

**Comment by dnil** [comment]
> > No, its for `scout update diseases` only. The file format change only dislodged the disease info from two of the files, the rest is still there.
> 
> AHH OK, it's confusing with all these files involved!

Agreed!

File: scout/build/disease.py [hunk 7]
```diff
@@ -70,8 +70,11 @@ def build_disease_term(disease_info, alias_genes={}):
                 for hgnc_id in alias_genes[hgnc_symbol]["ids"]:
                     hgnc_ids.add(hgnc_id)
         else:
-            LOG.debug("Gene symbol %s could not be found in database", hgnc_symbol)
+            hgnc_symbols_not_found.add(hgnc_symbol)
 
+    LOG.debug(
```
Comment (northwestwitch) [inline scout/build/disease.py:75]: If you add `if hgnc_symbols_not_found:` it becomes more complex than 15?

File: scout/load/hpo.py [hunk 8]
```diff
@@ -147,23 +179,64 @@ def load_disease_terms(adapter, genemap_lines, genes=None, hpo_disease_lines=Non
...
+
+
+def _parse_disease_term_info(
+    disease_info: Dict,
+    disease_annotations: Dict[str, Any],
+    disease_id: str,
+    hpo_term_to_symbol: Dict[Any, set],
+) -> Dict:
+    """
+    Startign from the OMIM disease terms (genemap2), update with HPO terms from
```
Comment (northwestwitch) [inline scout/load/hpo.py:216]:
```suggestion
    Starting from the OMIM disease terms (genemap2), update with HPO terms from
```

**Review by northwestwitch — APPROVED** [review]
Looks good!

File: scout/build/disease.py (hunk 7 above)
Comment (dnil) [reply scout/build/disease.py:75]: naaaah, lets.

File: scout/commands/setup/setup_scout.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dnil) [reply scout/commands/setup/setup_scout.py:96]: Ok, added that to the old Major issue.

**Comment by dnil** [comment]
Tested on stage:
```
...
2023-04-25 15:37:28 hasta.scilifelab.se scout.build.disease[26991] DEBUG The following gene symbols could not be found in database: {'DELYq11'}
2023-04-25 15:37:28 hasta.scilifelab.se scout.build.disease[26991] DEBUG Building disease term OMIM:425500
2023-04-25 15:37:28 hasta.scilifelab.se scout.build.disease[26991] DEBUG The following gene symbols could not be found in database: {'HEY'}
2023-04-25 15:37:28 hasta.scilifelab.se scout.build.disease[26991] DEBUG Building disease term OMIM:415000
2023-04-25 15:37:28 hasta.scilifelab.se scout.build.disease[26991] DEBUG Building disease term OMIM:400043
2023-04-25 15:37:28 hasta.scilifelab.se scout.build.disease[26991] DEBUG The following gene symbols could not be found in database: {'DFNY1'}
2023-04-25 15:37:28 hasta.scilifelab.se scout.build.disease[26991] DEBUG Building disease term OMIM:400004
2023-04-25 15:37:28 hasta.scilifelab.se scout.build.disease[26991] DEBUG The following gene symbols could not be found in database: {'RPY'}
Loading disease terms  [####################################]  100%
2023-04-25 15:37:28 hasta.scilifelab.se scout.load.hpo[26991] INFO Loading done. Nr of diseases loaded 7470
2023-04-25 15:37:28 hasta.scilifelab.se scout.load.hpo[26991] INFO Time to load diseases: 0:00:06.561359
2023-04-25 15:37:28 hasta.scilifelab.se scout.commands.update.disease[26991] INFO Successfully loaded all disease terms
```

**Comment by dnil** [comment]
Ok, I feel very done with this now. If there are no protests, I will merge. Tests are ok so far, and diseases look good both in number and when examined on the WUI. There could very much still be edge cases, or, say, things that show only when loading cases or using certain functionalities with the diseases.  We are looking at tests for the release as soon as anyway so why not test for any weird stuff with the entire release branch.
